File: src/backend/models/advertisementModel.js
const mongoose = require('mongoose');
const {Schema} = mongoose;

const priceTierSchema = new Schema({
    minQuantity: { type: Number, required: true },
    maxQuantity: { type: Number, required: true },
    price: { type: Number, required: true },
    unit: { type: String, required: true, enum: ['kg', 'g', 'L', 'ml', 'unit', 'units'] }
}, { _id: false });

const advertisementSchema = new Schema({
  seller: {
    type: Schema.Types.ObjectId,
    ref: 'Seller',
    required: true
  },
  category: {
    type: Schema.Types.ObjectId,
    ref: 'SellerCategory',
    required: true
  },
  subCategory: {
    type: Schema.Types.ObjectId,
    ref: 'SubCategory',
    required: true,
    validate: {
      validator: async function(subCategoryId) {
        // Validate that subcategory belongs to seller's category
        const subCategory = await mongoose.model('SubCategory').findById(subCategoryId);
        if (!subCategory) return false;
        
        const seller = await mongoose.model('Seller').findById(this.seller);
        return seller.category.equals(subCategory.category);
      },
      message: 'Subcategory must belong to seller\'s registered category'
    }
  },
  name: {
    type: String,
    required: [true, 'Advertisement name is required'],
    trim: true,
    maxlength: [100, 'Advertisement name cannot exceed 100 characters']
  },
  description: {
    type: String,
    required: [true, 'Description is required'],
    maxlength: [500, 'Description cannot exceed 500 characters']
  },
  quantity: {
    type: Number,
    required: [true, 'Quantity is required'],
    min: [0, 'Quantity cannot be negative']
  },
  unit: {
    type: String,
    required: true,
    enum: ['kg', 'g', 'L', 'ml', 'unit', 'units']
  },
  deliveryAvailable: {
    type: Boolean,
    required: true,
    default: false
  },
  deliveryRadius: {
    type: Number,
    required: function() { return this.deliveryAvailable; },
    min: [0, 'Delivery radius cannot be negative']
  },
  location: {
    type: String,
    required: [true, 'Location is required']
  },
  priceTiers: {
    type: [priceTierSchema],
    required: true,
    validate: {
      validator: function(tiers) {
        // Ensure no overlapping quantity ranges
        for (let i = 0; i < tiers.length; i++) {
          for (let j = i + 1; j < tiers.length; j++) {
            if (Math.max(tiers[i].minQuantity, tiers[j].minQuantity) <= 
                Math.min(tiers[i].maxQuantity, tiers[j].maxQuantity)) {
              return false;
            }
          }
        }
        return tiers.length > 0;
      },
      message: 'Price tiers have overlapping quantities or no tiers provided'
    }
  },
  images: [{
    type: String,
    validate: {
      validator: function(v) {
        return /\.(jpe?g|png|gif|bmp)$/i.test(v);
      },
      message: props => `${props.value} is not a valid image file!`
    }
  }],
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, { timestamps: true });

// Middleware to validate seller's category matches the product category
advertisementSchema.pre('save', async function(next) {
  const Seller = mongoose.model('Seller');
  const seller = await Seller.findById(this.seller).populate('category');
  
  if (!seller) {
    throw new Error('Invalid seller reference');
  }
  
  if (!seller.category._id.equals(this.category)) {
    throw new Error('Product category must match seller\'s registered category');
  }
  
  next();
});

// Indexes for better query performance
advertisementSchema.index({ seller: 1 });
advertisementSchema.index({ category: 1 });
advertisementSchema.index({ subCategory: 1 });
advertisementSchema.index({ isActive: 1 });
advertisementSchema.index({ location: 'text', name: 'text', description: 'text' });

const Advertisement = mongoose.model('Advertisement', advertisementSchema);

module.exports = Advertisement;

File: src/frontend/lib/models/advertisement_model.dart
import 'package:frontend/models/tender_model.dart';

class PriceTier {
  final double minQuantity;
  final double maxQuantity;
  final double price;
  final Units unit;

  PriceTier({
    required this.minQuantity,
    required this.maxQuantity,
    required this.price,
    required this.unit,
  });

  factory PriceTier.fromJson(Map<String, dynamic> json) {
    return PriceTier(
      minQuantity: json['minQuantity'].toDouble(),
      maxQuantity: json['maxQuantity'].toDouble(),
      price: json['price'].toDouble(),
      unit: unitsFromString(json['unit']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'minQuantity': minQuantity,
      'maxQuantity': maxQuantity,
      'price': price,
      'unit': unitsToString(unit),
    };
  }
}

class AdvertisementModel {
  String? id;
  String? sellerId;
  String? categoryId;
  final String subCategoryId;
  final String name;
  final String description;
  final double quantity;
  final Units unit;
  final bool deliveryAvailable;
  final double? deliveryRadius;
  String? location;
  final List<PriceTier> priceTiers;
  List<String>? images;
  DateTime? createdAt;
  DateTime? updatedAt;
  bool? isActive;

  AdvertisementModel({
    this.id,
    this.sellerId,
    this.categoryId,
    required this.subCategoryId,
    required this.name,
    required this.description,
    required this.quantity,
    required this.unit,
    required this.deliveryAvailable,
    this.deliveryRadius,
    this.location,
    required this.priceTiers,
    this.images,
    this.createdAt,
    this.updatedAt,
    this.isActive,
  });

  factory AdvertisementModel.fromJson(Map<String, dynamic> json) {
    return AdvertisementModel(
      id: json['_id'],
      sellerId: json['seller'],
      categoryId: json['category'],
      subCategoryId: json['subCategory'],
      name: json['name'],
      description: json['description'],
      quantity: json['quantity'].toDouble(),
      unit: unitsFromString(json['unit']),
      deliveryAvailable: json['deliveryAvailable'],
      deliveryRadius: json['deliveryRadius']?.toDouble(),
      location: json['location'],
      priceTiers: List<PriceTier>.from(
        json['priceTiers'].map((x) => PriceTier.fromJson(x)),
      ),
      images: List<String>.from(json['images']),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      isActive: json['isActive'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'subCategory': subCategoryId,
      'name': name,
      'description': description,
      'quantity': quantity,
      'unit': unitsToString(unit),
      'deliveryAvailable': deliveryAvailable,
      'deliveryRadius': deliveryRadius,
      'priceTiers': priceTiers.map((x) => x.toJson()).toList(),
      'images': images,
    };
  }
}
